app/services/feeds/optimized.rb

```diff
@@ -0,0 +1,279 @@
+        article_points = score_single_article(article)
+        result[article] = article_points
+      end
+      ranked_articles = ranked_articles.sort_by { |_article, article_points| -article_points }.map(&:first)
+      ranked_articles.to(@number_of_articles - 1)
+    end
+
+    def score_single_article(article)
+      article_points = 0
+      article_points += score_followed_user(article)
+      article_points += score_followed_organization(article)
+      article_points += score_followed_tags(article)
+      article_points += score_randomness
+      article_points += score_language(article)
+      article_points += score_experience_level(article)
+      article_points += score_comments(article)
+      article_points
+    end
+
+    def score_followed_user(article)
+      user_following_users_ids.include?(article.user_id) ? 1 : 0
+    end
+
+    def score_followed_tags(article)
+      return 0 unless @user
+
+      article_tags = article.decorate.cached_tag_list_array
+      user_followed_tags.sum do |tag|
+        article_tags.include?(tag.name) ? tag.points * @tag_weight : 0
+      end
+    end
+
+    def score_followed_organization(article)
+      user_following_org_ids.include?(article.organization_id) ? 1 : 0
+    end
+
+    def score_randomness
+      rand(3) * @randomness
+    end
+
+    def score_language(article)
+      @user&.preferred_languages_array&.include?(article.language || "en") ? 1 : -15
+    end
+
+    def score_experience_level(article)
+      - (((article.experience_level_rating - (@user&.experience_level || 5)).abs / 2) * @experience_level_weight)
+    end
+
+    def score_comments(article)
+      article.comments_count * @comment_weight
+    end
+
+    def globally_hot_articles(user_signed_in)
+      hot_stories = published_articles_by_tag
+        .where("score > ? OR featured = ?", 7, true)
+        .order(hotness_score: :desc)
+      featured_story = hot_stories.where.not(main_image: nil).first
+      if user_signed_in
+        hot_story_count = hot_stories.count
+        offset = RANDOM_OFFSET_VALUES.select do |i|
+          i < hot_story_count
+        end.sample # random offset, weighted more towards zero
+        hot_stories = hot_stories.offset(offset)
+        new_stories = Article.published
+          .where("score > ?", -15)
+          .limited_column_select.includes(top_comments: :user).order(published_at: :desc).limit(rand(15..80))
+        hot_stories = hot_stories.to_a + new_stories.to_a
+      end
+      [featured_story, hot_stories.to_a]
+    end
+
+    private
+
+    def user_followed_tags
+      @user_followed_tags ||= (@user&.decorate&.cached_followed_tags || [])
+    end
+
+    def user_following_org_ids
+      @user_following_org_ids ||= (@user&.cached_following_organizations_ids || [])
+    end
+
+    def user_following_users_ids
+      @user_following_users_ids ||= (@user&.cached_following_users_ids || [])
+    end
+
+    def first_half(array)
+      array[0...(array.length / 2)]
```

Comment:
We can lean a bit more on Ruby here:

```suggestion
      array.first(array.length / 2)
```